File: implementations/csharp/server/tests/ShuaiTunnel.IntegrationTests/PublicTransferClusterFrameTests.cs
using System.Text;
using System.Text.Json;
using ShuaiTunnel.Server.WebSockets;

namespace ShuaiTunnel.IntegrationTests;

public sealed class PublicTransferClusterFrameTests
{
    [Fact]
    public void TextEventMatchesCanonicalVectorAndRoundTrips()
    {
        using var document = JsonDocument.Parse(File.ReadAllText(FindVector()));
        var root = document.RootElement;
        var vector = root.GetProperty("canonicalText");
        var clusterEvent = new PublicTransferClusterEvent(
            vector.GetProperty("kind").GetByte(),
            vector.GetProperty("excludeSource").GetBoolean(),
            vector.GetProperty("revision").GetUInt64(),
            vector.GetProperty("groupId").GetString()!,
            vector.GetProperty("targetPeerId").GetString()!,
            vector.GetProperty("sourceLeaseId").GetString()!,
            Encoding.UTF8.GetBytes(vector.GetProperty("payloadUtf8").GetString()!));

        var encoded = PublicTransferClusterFrame.Encode(clusterEvent);
        Assert.Equal(Convert.FromHexString(vector.GetProperty("frameHex").GetString()!), encoded);

        var decoded = PublicTransferClusterFrame.Decode(encoded);
        Assert.Equal(clusterEvent.Kind, decoded.Kind);
        Assert.Equal(clusterEvent.ExcludeSource, decoded.ExcludeSource);
        Assert.Equal(clusterEvent.Revision, decoded.Revision);
        Assert.Equal(clusterEvent.GroupId, decoded.GroupId);
        Assert.Equal(clusterEvent.TargetPeerId, decoded.TargetPeerId);
        Assert.Equal(clusterEvent.SourceLeaseId, decoded.SourceLeaseId);
        Assert.Equal(clusterEvent.Payload, decoded.Payload);
        Assert.Throws<ArgumentException>(() =>
            PublicTransferClusterFrame.Decode([.. encoded, 0]));

        var derivation = root.GetProperty("groupIdDerivation");
        Assert.Equal(derivation.GetProperty("groupId").GetString(),
            PublicTransferCoordinationService.GroupId(
                derivation.GetProperty("roomId").GetString()!,
                derivation.GetProperty("roomKey").GetString()!));
    }

    [Fact]
    public void BinaryWithoutTargetAndRosterPayloadAreRejected()
    {
        Assert.Throws<ArgumentException>(() => PublicTransferClusterFrame.Encode(
            new PublicTransferClusterEvent(PublicTransferClusterFrame.KindBinary, false, 0,
                "group", string.Empty, string.Empty, [1])));
        Assert.Throws<ArgumentException>(() => PublicTransferClusterFrame.Encode(
            new PublicTransferClusterEvent(PublicTransferClusterFrame.KindRoster, false, 1,
                "group", string.Empty, string.Empty, [1])));
    }

    [Fact]
    public void ManagementEventMatchesCanonicalVectorAndTenantBinding()
    {
        using var document = JsonDocument.Parse(File.ReadAllText(FindVector()));
        var vector = document.RootElement.GetProperty("canonicalManagement");
        var clusterEvent = new PublicTransferClusterEvent(
            vector.GetProperty("kind").GetByte(),
            vector.GetProperty("excludeSource").GetBoolean(),
            vector.GetProperty("revision").GetUInt64(),
            vector.GetProperty("groupId").GetString()!,
            vector.GetProperty("targetPeerId").GetString()!,
            vector.GetProperty("sourceLeaseId").GetString()!,
            Encoding.UTF8.GetBytes(vector.GetProperty("payloadUtf8").GetString()!));

        var encoded = PublicTransferClusterFrame.Encode(clusterEvent);

        Assert.Equal(Convert.FromHexString(vector.GetProperty("frameHex").GetString()!), encoded);
        var decoded = PublicTransferClusterFrame.Decode(encoded);
        Assert.Equal(clusterEvent.Kind, decoded.Kind);
        Assert.Equal(clusterEvent.GroupId, decoded.GroupId);
        Assert.Equal(clusterEvent.Payload, decoded.Payload);
        Assert.Equal(vector.GetProperty("groupId").GetString(),
            PublicTransferCoordinationService.ManagementGroupId(
                vector.GetProperty("tenantId").GetString()));
        Assert.Throws<ArgumentException>(() => PublicTransferClusterFrame.Encode(
            clusterEvent with { TargetPeerId = "unexpected-target" }));
    }

    private static string FindVector()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        for (var depth = 0; directory is not null && depth < 12; depth++, directory = directory.Parent)
        {
            var candidate = Path.Combine(directory.FullName, "protocol", "test-vectors",
                "public-transfer-cluster-v2.json");
            if (File.Exists(candidate))
            {
                return candidate;
            }
        }
        throw new FileNotFoundException("cannot locate public-transfer-cluster-v2.json");
    }
}

File: implementations/java/server/src/test/java/com/theshuai/specusserver/websocket/PublicTransferClusterFrameTests.java
package com.theshuai.specusserver.websocket;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.HexFormat;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PublicTransferClusterFrameTests {

    @Test
    void textEventMatchesCanonicalVectorAndRoundTrips() throws Exception {
        JsonNode root = new ObjectMapper().readTree(Files.readString(findVector()));
        JsonNode vector = root.path("canonicalText");
        byte[] payload = vector.path("payloadUtf8").asText()
                .getBytes(java.nio.charset.StandardCharsets.UTF_8);
        PublicTransferClusterFrame.Event event = new PublicTransferClusterFrame.Event(
                (byte) vector.path("kind").asInt(),
                vector.path("excludeSource").asBoolean(),
                vector.path("revision").asLong(),
                vector.path("groupId").asText(),
                vector.path("targetPeerId").asText(),
                vector.path("sourceLeaseId").asText(),
                payload);
        byte[] encoded = PublicTransferClusterFrame.encode(event);

        assertThat(encoded).containsExactly(HexFormat.of().parseHex(vector.path("frameHex").asText()));

        PublicTransferClusterFrame.Event decoded = PublicTransferClusterFrame.decode(
                encoded);

        assertThat(decoded.kind()).isEqualTo(PublicTransferClusterFrame.KIND_TEXT);
        assertThat(decoded.excludeSource()).isTrue();
        assertThat(decoded.revision()).isEqualTo(17);
        assertThat(decoded.groupId()).isEqualTo("0123456789abcdef");
        assertThat(decoded.targetPeerId()).isEqualTo("peer-b");
        assertThat(decoded.sourceLeaseId()).isEqualTo("lease-a");
        assertThat(decoded.payload()).containsExactly(payload);
    }

    @Test
    void managementEventMatchesCanonicalVectorAndTenantBinding() throws Exception {
        JsonNode vector = new ObjectMapper().readTree(Files.readString(findVector()))
                .path("canonicalManagement");
        byte[] payload = vector.path("payloadUtf8").asText()
                .getBytes(java.nio.charset.StandardCharsets.UTF_8);
        PublicTransferClusterFrame.Event event = new PublicTransferClusterFrame.Event(
                (byte) vector.path("kind").asInt(),
                vector.path("excludeSource").asBoolean(),
                vector.path("revision").asLong(),
                vector.path("groupId").asText(),
                vector.path("targetPeerId").asText(),
                vector.path("sourceLeaseId").asText(),
                payload);

        byte[] encoded = PublicTransferClusterFrame.encode(event);

        assertThat(encoded).containsExactly(HexFormat.of().parseHex(vector.path("frameHex").asText()));
        PublicTransferClusterFrame.Event decoded = PublicTransferClusterFrame.decode(encoded);
        assertThat(decoded.kind()).isEqualTo(event.kind());
        assertThat(decoded.groupId()).isEqualTo(event.groupId());
        assertThat(decoded.payload()).containsExactly(payload);
        assertThat(PublicTransferCoordinationService.managementGroupId(vector.path("tenantId").asText()))
                .isEqualTo(vector.path("groupId").asText());
        assertThatThrownBy(() -> PublicTransferClusterFrame.encode(
                new PublicTransferClusterFrame.Event(
                        PublicTransferClusterFrame.KIND_MANAGEMENT, false, 0,
                        event.groupId(), "unexpected-target", "", payload)))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("management event shape");
    }

    @Test
    void decoderRejectsTrailingBytesAndBinaryWithoutTarget() {
        PublicTransferClusterFrame.Event event = new PublicTransferClusterFrame.Event(
                PublicTransferClusterFrame.KIND_ROSTER,
                false,
                1,
                "group",
                "",
                "",
                new byte[0]);
        byte[] encoded = PublicTransferClusterFrame.encode(event);
        assertThatThrownBy(() -> PublicTransferClusterFrame.decode(
                Arrays.copyOf(encoded, encoded.length + 1)))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("length mismatch");

        assertThatThrownBy(() -> PublicTransferClusterFrame.encode(
                new PublicTransferClusterFrame.Event(
                        PublicTransferClusterFrame.KIND_BINARY,
                        false,
                        0,
                        "group",
                        "",
                        "",
                        new byte[]{1})))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void groupIdIsStableAndSeparatesRoomComponents() throws Exception {
        JsonNode derivation = new ObjectMapper().readTree(Files.readString(findVector()))
                .path("groupIdDerivation");
        assertThat(PublicTransferCoordinationService.groupId("nearby", "public:203.0.113.8"))
                .isEqualTo(derivation.path("groupId").asText());
        assertThat(PublicTransferCoordinationService.groupId("a\nb", "c"))
                .isNotEqualTo(PublicTransferCoordinationService.groupId("a", "b\nc"));
    }

    private static Path findVector() {
        Path current = Path.of("").toAbsolutePath();
        for (int depth = 0; current != null && depth < 8; depth++, current = current.getParent()) {
            Path candidate = current.resolve("protocol/test-vectors/public-transfer-cluster-v2.json");
            if (Files.isRegularFile(candidate)) {
                return candidate;
            }
        }
        throw new IllegalStateException("cannot locate public transfer cluster v2 vector");
    }
}
